### ISSUE TITLE
2.0 RC3 fresh install not returning focus after exiting a game

### ISSUE THREAD

**Issue by chris-vg** [mickelson/attract#199]
Problem:
After launching a game via AM, exit the game normally and AM stays in it's sleep state (appears frozen, cannot interact with it) and you can see the mouse cursor displayed.

Work around:
press Alt-Tab and AM wakes up and can be used again like normal.

Steps to reproduce:
Create a fresh installation of AM RC3 (or the current nightly RC3-1)
Select English
Select Emulators
Select mame
Set up mame exe path
Set up mame rom path
Generate list
Exit config
Launch a game from the list
Exit mame by pressing ESC

I can reproduce this on 2 machines, both have Windows 10.

Also, I have tested this scenario with earlier versions of AM (1.62 and 2.0 RC2) and this problem does not occur.

Finally, as an aside, I have another copy of RC3 that i've been using as my main setup and it does not exhibit this behavior on one of the machines, but it does on the other.

**Comment by mickelson** [comment]
Hey Chris, what version of mame are you using?

**Comment by chris-vg** [comment]
Sorry, I didn't even think to mention it.  I am having this problem with 0.171 GroovyMAME.

I just performed a test with 0.171 standard MAMEDev release and the problem is not there.

It's strange that I only have this problem with RC3 though, I have no problems returning to AM RC2 with GroovyMAME.

**Comment by h0tw1r3** [comment]
@chris-vg @omegaman1 Could you test this build please? https://build.zaplabs.com/dl/project/attractmode/attract64-v2.0.0-rc3-1-windows.zip

@mickelson Upgraded build server gcc from 4.9.8 to 5.3.0.

**Comment by omegaman1** [comment]
Tried the new build. Sorry, it didn't fix the problem for me.

**Comment by chris-vg** [comment]
Tested on both my desktop and blast city cabinet.

Works fine on my desktop with groovymame, exit game and returns to AM nicely.
Problem still occurs on my arcade cabinet with groovymame, exit game and need to alt tab to make AM responsive again.

One thing that may explain the difference I'm seeing between both machines is that groovymame on my cab uses the built in switchres code to set the correct video resolution and refresh rate for each game.  on my desktop i bypass switchres so it always runs at my desktop native res.

I've attached dxdiags for both of my machines, I don't think there will be anything there that will help to resolve the issue, but you never know...

[DxDiag-BlastCity.txt](https://github.com/mickelson/attract/files/183510/DxDiag-BlastCity.txt)
[DxDiag-Desktop.txt](https://github.com/mickelson/attract/files/183511/DxDiag-Desktop.txt)

**Comment by chris-vg** [comment]
Quick update, disabling switchres on my arcade cab did not make a difference, so ignore that i guess.

**Comment by omegaman1** [comment]
What happens when you play a game for more than 5 minutes? That's about how long it takes for it to become unresponsive for me. I created a dump file but its too big to post. I need to download the ms sdk kit for debugging tools like Dumpchk...

**Comment by chris-vg** [comment]
@omegaman1 I'm not so sure we are experiencing the same issue.  I don't get any crashes at all, I just  find that when exiting a game, AM is not returning to focus, so I need to Alt+Tab to wake it up and then I can continue to use the frontend like normal.

I did test running a game on both of my machines for well over 5 minutes a piece and no crashes at all.

**Comment by omegaman1** [comment]
Yeah, I understand that. That's why I don't think my issue is a focus problem. I'm just at a complete loss as to why am becomes unresponsive when exiting the emulator if I play for say 5 minutes. Time to start looking at updates or some out the box type stuff. On windows 7 and Linux everything is working great.

**Comment by CapnBry** [comment]
I am experiencing the same issue. Launch into MAME and when MAME exits, Attract-Mode does not get focus back or something and I have to alt-tab to it. I am also on Windows 10, but MAME 161. I've tried enable/disable multi monitors in AM, changed from Fullscreen to Fill, neither seemed to make a difference. I tried enabling the resfix plugin as well. I wrote a little program to show me what window has focus at all times and it appears `GetClassName(GetForegroundWindow())` returns "MAME" but `GetWindowText()` returns blank (while in MAME GetWindowText returns what you'd expect).

It does seem that Attract-Mode is not getting focus back after MAME is exiting for some reason. The updated build above from @h0tw1r3 has the same results. Perhaps AM needs to bring itself to the foreground after the game exits?

I modified my app to SetForegroundWindow to the Attract-Mode window if it sees that MAME with no window text is the foreground window. This **works like a champ**, but because I poll the foreground window it sometimes takes a couple tries to bring AM back to the foreground. I use this delphi code because it is the only native Win32 development environment I have on hand currently, but this is easy enough to port to C where the hWndToForce is the AM window. Perhaps placed [here](https://github.com/mickelson/attract/blob/master/src/fe_util.cpp#L862)
```
function W2K_SetForegroundWindow(hWndToForce : HWND) : Boolean;
var
   hCurrWnd:  HWND;
   dwMyTID:   DWORD;
   dwCurrTID: DWORD;
begin
   hCurrWnd := GetForegroundWindow;
   if hCurrWnd = hWndToForce then begin
    Result := true;
    exit;
   end;

   dwMyTID   := GetCurrentThreadId;
   dwCurrTID := GetWindowThreadProcessId(hCurrWnd, nil);

   AttachThreadInput(dwMyTID, dwCurrTID, True);
    { Now we look like the foreground process, we can set the foreground window. }
   Result := SetForegroundWindow (hWndToForce);
   AttachThreadInput(dwMyTID, dwCurrTID, False);
end;
```

**Comment by omegaman1** [comment]
I found this nugget about win10 - some devs discussing how window is not put to the front on focus.  

If I call window.minimize() right before window.focus(), I can get the window to the foreground no matter what on Windows 10.

https://github.com/atom/electron/issues/2867

**Comment by h0tw1r3** [comment]
@CapnBry @chris-vg Can you test this build? https://build.zaplabs.com/dl/project/attractmode/.temp/attract-focus-test1.zip

This build fixes the problem for me. I'm running in desktop mode, and launching from explorer.

Are you running Windows 10 in Desktop or Tablet mode?
How are you launching AM (console, script, explorer/icon)?

**Comment by CapnBry** [comment]
That build didn't change anything for me. I am launching AM from Windows 10 Desktop mode by double clicking the icon in an Explorer window (not from Start Menu or Desktop icon).

I grabbed the source code and built MXE (whew that took a while!) and this patch works for me:
```
diff --git a/src/fe_window.cpp b/src/fe_window.cpp
index ed3ac25..a88d94f 100644
--- a/src/fe_window.cpp
+++ b/src/fe_window.cpp
@@ -254,6 +254,9 @@ bool FeWindow::run()
        //
        m_fes.run( min_run );

+       sf::WindowHandle hw = getSystemHandle();
+       SetForegroundWindow(hw);
+
 #ifdef SFML_SYSTEM_MACOS
        osx_take_focus();
 #endif
```

Basically the same place MacOS grabs focus back, pull back focus on Windows. I didn't add the SFML_SYSTEM_WINDOWS conditional but obviously you'll want that. Note that if for some reason AM is not the foreground *process* already then it will just blink the taskbar item instead of switching. This can be subverted by the code I have in a previous post with the AttachThreadInput code, but it does not appear to be necessary.

**Comment by h0tw1r3** [comment]
@CapnBry With SFML, this cross-platform method _should_ work:
```
diff --git a/src/fe_window.cpp b/src/fe_window.cpp
index ed3ac25..a88d94f 100644
--- a/src/fe_window.cpp
+++ b/src/fe_window.cpp
@@ -254,6 +254,8 @@ bool FeWindow::run()
        //
        m_fes.run( min_run );

+       requestFocus();
+
 #ifdef SFML_SYSTEM_MACOS
        osx_take_focus();
 #endif
```

**Comment by CapnBry** [comment]
Oh that would be great, however it does not work :(

Looking at the SFML source code it will only SetForegroundWindow if it is the foreground process (the foreground window's thread processID is AM). Apparently, AM is not the foreground process because it just blinks the window. It is like there is still a MAME window somewhere which only exists until AM gets focus again. It is very bizarre. I looked through the code thinking that maybe there's still a handle open to some part of the child process which is keeping the process around until AM gets focus again but it all looks clean to me. Even ProcExp is showing the MAME child process as exited.

Clearly *nothing* has focus though because Windows wouldn't let you grab foreground with just SetForegroundWindow() if another process had input focus. It does let you though, which indicates that it is perfectly OK?

So TL/DR: requestFocus doesn't get us where we need to be.

**Comment by h0tw1r3** [comment]
@mickelson Reverting 8e775a8969ad9e1abd733eaa5777032673a1d86d and 5bf8f1061ac201ded1b0c8fd2544c9c5c2a20429 (console hiding) fixes this issue for me (and #192 also).

Reading up on this general problem (unrelated to AM, SFML ,etc) there are a lot of handle issues when running a console program with a gui and trying to manage CreateProcess programs. Particularly when those programs can be console w/gui like mame. I see in the AM code, it's already show up (WP_POPUP work-around).

Perhaps the best way to eliminate the problem is to link AM as a real GUI and wedge in the command line support. It seems to be the only possible way of doing this without playing patch-a-mole (and fighting against the way SFML does things).

Good read: https://github.com/rprichard/win32-console-docs/blob/master/README.md

**Comment by chris-vg** [comment]
@h0tw1r3 I don't have a build environment set up here, any chance I could get a copy of your build to try out here?

**Comment by h0tw1r3** [comment]
@chris-vg big fat wip: https://build.zaplabs.com/dl/project/attractmode/.temp/attract-issue199-gui.zip

**Comment by CapnBry** [comment]
@chris-vg Here's my build as well with the SetForegroundWindow workaround. If the 199 version doesn't work for you (doesn't make a difference for me) try this one and let us know https://home.capnbry.net:22676/~bmayland/attract-199.zip

**Comment by mickelson** [comment]
holy crap!: https://github.com/rprichard/win32-console-docs/blob/master/README.md

good find h0tw1r3, but what a pain...

**Comment by CapnBry** [comment]
I can confirm that the latest master branch code works perfectly for me now. Thanks!

Might want to wait until the next proper RC so others can confirm before closing the ticket though?

**Comment by chris-vg** [comment]
I've just quickly tested the latest nightly and this problem does indeed appear to be resolved for me too.

**Comment by chris-vg** [comment]
Tested the 2.0 release build some more, and while the problem seems to be fixed on my desktop machine (win10), for some odd reason the problem is still happening on my arcade machine (win10 as well).

I will need to do some more tests to try and identify what difference between the machines is causing this to occur.

**Comment by mickelson** [comment]
Ok, holding this issue open to hear back about chris' tests.

**Comment by mickelson** [comment]
@chris-vg - Are you still having this issue on your arcade machine with 2.0?

**Comment by chris-vg** [comment]
Sorry for not getting back to you on to this sooner.

Yes, I am still experiencing this issue on my arcade box with AM v2.0 release build.

**Comment by omegaman1** [comment]
I'm having the same problem again on win 10 when I exit out of mame. The goods news is the windows key unfreezes AM, unlike before. Oddly, the problem seems to go away after the initial freeze and ctrl+escape is executed. 

This is a deal breaker for windows 10 users.

**Comment by mickelson** [comment]
Related problem... Configuring the "exit hotkey" in Attract-Mode to the same key that is configured in Mame to exit will cause AM to lose focus on return:

https://github.com/mickelson/attract/issues/244
